Added +lsp to Tex-Mode.

MYDavoodeh opened the pull request.

yoavm448 commented:
Doesn't this also need `:tools lsp` to work? You didn't check for `:tools lsp`anywhere, nor is it mentioned docs.

Also, doom doesn't install yatex, why is a hook added to it? If users installed yatex themselves, let them add the hook as they wish.

With that out of the way, how is lsp with latex? Does it provide features that AUCTeX is missing? I tried and failed to download digestif a while back, so I don't know.

P.S. Also document the `+lsp` flag on `docs/intex.org`

MYDavoodeh commented:
Yes all of those above are true... The thing is I'm just a bit new to this...
Thanks for addressing the issues...
No, I'm not sure if AUCTeX is any less than this. But anyways almost every language has an LSP flag so why not Latex. Especially now (about a week now) that TexLab is also compatible and it's not only about the limited "digestif".
Trying to fix the addressed issues.

Feel free to close this if it seems obsolete! It is just a suggestion :D
Thanks for your kind issue addressing

yoavm448 commented:
Ooooh I didn't know that lsp now supports TeX-lab, I'll try it!
I didn't mean to be offensive, just commented on a few things that bugged me about the code. I only know about some of those things because I made these mistakes myself. I'm happy that the latex module gets more love! Hopefully I'll PR a +fold flag for it soon, once I have it polished.

hlissner commented on modules/lang/latex/config.el:
```diff
@@ -180,6 +180,10 @@ If no viewers are found, `latex-preview-pane' is used.")
+(after! tex-mode
+  (when (and (featurep! +lsp) (featurep! :tools lsp))
+    (add-hook 'tex-mode-local-vars-hook #'lsp!)
+    (add-hook 'latex-mode-local-vars-hook #'lsp!)))
```
Could you move this into the `(after! tex ...)` block further up? To reduce the number of "floating" blocks of related code.

MYDavoodeh replied:
> Side note: should we not mention the LSP server meant to be installed for latex? I see no mention of it in `+lsp`'s description, or anywhere in the README.

I thought about it actually and I saw no point in it. Since it's greatly explained in the LSP-Mode
and it's just about 2 lines of code in the `clients.el`. Moreover, once you don't have the clients and run LSP, it will warn and instruct you to install servers... So I'm not sure!

If it is required, I will add it.

MYDavoodeh commented:
> Side note: should we not mention the LSP server meant to be installed for latex? I see no mention of it in `+lsp`'s description, or anywhere in the README.

I thought about it actually and I saw no point in it. Since it's greatly explained in the LSP-Mode
and it's just about 2 lines of code in the `clients.el`. Moreover, once you don't have the clients and run LSP, it will warn and instruct you to install servers... So I'm not sure!

If it is required, I will add it.

hlissner commented:
> Moreover, once you don't have the clients and run LSP, it will warn and instruct you to install servers...

[I have disabled LSP's ability to auto-install servers](https://github.com/hlissner/doom-emacs/blob/develop/modules/tools/lsp/config.el#L52-L57) believe it is poor etiquette. If users want to install a server they can do so deliberately with `M-x lsp-install-server` or by installing/building the servers themselves (generally through their system package manager).

As such it is better you mention it and link to it in the README. Doesn't have to be more than a name drop.

hlissner commented on docs/modules.org:
```diff
@@ -114,7 +114,7 @@ Modules that bring support for a language or group of languages to Emacs.
 + julia - TODO
 + kotlin =+lsp+= - TODO
-+ [[file:../modules/lang/latex/README.org][latex]] =+latexmk +cdlatex= - TODO
++ [[file:../modules/lang/latex/README.org][latex]] =+latexmk +cdlatex= =+lsp= - TODO
```
There are extra `=`'s here. e.g.

`=+latexmk +cdlatex= =+lsp=` should be `=+latexmk +cdlatex +lsp=`

hlissner commented on modules/lang/latex/config.el:
```diff
@@ -74,7 +74,11 @@ If no viewers are found, `latex-preview-pane' is used.")
-      (sp-local-pair modes "``" nil :unless '(:add sp-in-math-p)))))
+      (sp-local-pair modes "``" nil :unless '(:add sp-in-math-p))))
+  ;; Hook lsp if enabled
+  (when (and (featurep! +lsp) (featurep! :tools lsp))
```
The `(featurep! :tools lsp)` check isn't necessary and should be omitted, in case the user has set up LSP themselves.

hlissner commented on modules/lang/latex/config.el:
```diff
@@ -74,7 +74,11 @@ If no viewers are found, `latex-preview-pane' is used.")
+      (sp-local-pair modes "``" nil :unless '(:add sp-in-math-p))))
+  ;; Hook lsp if enabled
+  (when (and (featurep! +lsp) (featurep! :tools lsp))
+    (add-hook 'tex-mode-local-vars-hook #'lsp!)
+    (add-hook 'latex-mode-local-vars-hook #'lsp!)))
```
Please combine these with `add-hook!`:

```elisp
(add-hook! '(tex-mode-local-vars-hook
             latex-mode-local-vars-hook)
           #'lsp!)
```

hlissner commented on modules/lang/latex/config.el:
```diff
@@ -180,6 +184,5 @@ If no viewers are found, `latex-preview-pane' is used.")
   :init
   (add-to-list '+latex--company-backends #'+latex-symbols-company-backend nil #'eq))
 
```
Please restore this newline. Unrelated `use-package!` blocks should be separated by two blank lines.

tommy7420 replied:
find 5743771534

hlissner commented on modules/lang/latex/README.org:
```diff
@@ -30,6 +30,8 @@ Provide a helping hand when working with LaTeX documents.
 ** Module Flags
 + =+latexmk= Use LatexMk instead of LaTeX to compile documents.
 + =+cdlatex= Enable [[https://github.com/cdominik/cdlatex][cdlatex]] for fast math insertion.
++ =+lsp= Start LSP automatically in `tex-mode-hook`. This
+  requires the =:tools lsp= module.
```
Please reflow this with `org-fill-paragraph` (on `gQ`, if you use evil)

MYDavoodeh replied:
Thanks a lot! Didn't have an idea about that.

tecosaur commented:
I'd be in favour of adding `TeXlab` to `M-x lsp-install-server`, just my 2c.

carlosperezm commented:
I have been testing this PR (I am curious about `TexLab`) but unfortunately the completion is not working as intended.
The candidates are there but are not reordered based on the input. See screenshot below.

![output-2020-04-14-13:52:28](https://user-images.githubusercontent.com/7297323/79222622-2d45ea80-7e58-11ea-9ac9-696aa8c55e11.gif)

hlissner commented:
@carlosperezm That is an unrelated issue with company-lsp that I've fixed in https://github.com/hlissner/doom-emacs/commit/6315984ae642. Please update Doom.

hlissner commented:
Thanks for the help!
